analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,172 @@
+
+    private static string FindRoot(IDictionary<string, string> parents, string element) =>
+        parents[element] is var root && root != element ? FindRoot(parents, root) : root;
+
+    private static List<List<string>> DisjointSets(IDictionary<string, string> parents) =>
+        parents.GroupBy(x => FindRoot(parents, x.Key), x => x.Key).Select(x => x.OrderBy(x => x).ToList()).OrderBy(x => x[0]).ToList();
+
+    private static IEnumerable<Location> LocationIdentifiers<T>(ISymbol symbol) where T : SyntaxNode =>
+        symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<T>().Select(x => x.GetIdentifier()?.GetLocation());
+
+    private static ImmutableHashSet<string> RelevantMemberNames(INamedTypeSymbol symbol)
+    {
+        var builder = ImmutableHashSet.CreateBuilder<string>();
+        foreach (var member in symbol.GetMembers())
+        {
+            switch (member)
+            {
+                // Constructors are not considered because they have to be split anyway
+                // Accessors are not considered because they are part of properties, that are considered as a whole
+                case IMethodSymbol method when !method.IsConstructor() && !method.IsStaticConstructor() && method.AssociatedSymbol is not IPropertySymbol:
+                    builder.Add(method.Name);
+                    break;
+                // Primary constructor parameters may or may not generate fields, and must be considered
+                case IMethodSymbol method when method.IsPrimaryConstructor():
+                    builder.UnionWith(method.Parameters.Where(IsService).Select(x => x.Name));
+                    break;
+                case IFieldSymbol field when IsService(field) && !field.IsImplicitlyDeclared:
+                    builder.Add(field.Name);
+                    break;
+                case IPropertySymbol property when IsService(property):
+                    builder.Add(property.Name);
+                    break;
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    private static bool IsService(ISymbol symbol) =>
+        symbol.GetSymbolType() is { TypeKind: TypeKind.Interface, Name: var name } && !ExcludedWellKnownServices.Contains(name);
+
+    private static IEnumerable<SecondaryLocation> SecondaryLocations(INamedTypeSymbol controllerSymbol, List<List<string>> sets)
+    {
+        var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>();
+        return
+            from set in sets.Zip(Enumerable.Range(1, sets.Count), (set, setIndex) => new { Set = set, SetIndex = setIndex })
+            let setIndex = set.SetIndex
+            from memberName in set.Set
+            from method in methods.Where(x => x.Name == memberName)
+            from location in LocationIdentifiers<MethodDeclarationSyntax>(method)
+            select new SecondaryLocation(location, $"Belongs to responsibility #{setIndex}.");
+    }
```

Comment:
I think going for an old-fashioned double for-loop is both more readable and more performant here:

```csharp
var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>().ToList();
for (var setIndex = 0; setIndex < sets.Count; setIndex++)
{
    foreach (var memberLocation in sets[setIndex].SelectMany(MemberLocations))
    {
        yield return new SecondaryLocation(memberLocation, $"Belongs to responsibility #{setIndex}.");
    }
}

IEnumerable<Location> MemberLocations(string memberName) =>
    methods.Where(x => x.Name == memberName).SelectMany(LocationIdentifiers<MethodDeclarationSyntax>);
```

Advantages: 
- You do not need to make an enumerable range just to get the set index indirectly
- You do not need to zip, making more enumerators
- You do not need to use archaic query-syntax incantations (joking, it's a preference and sometimes stronger)
- Semantically, each loop does one of the two things you need:
  - Outer loop gives you setIndex
  - Inner loop gives you location